File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwServiceError.java
package org.wikipedia.dataclient.mwapi;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.dataclient.ServiceError;
import org.wikipedia.util.DateUtil;
import org.wikipedia.util.ThrowableUtil;

import java.util.Date;
import java.util.List;

/**
 * Gson POJO for a MediaWiki API error.
 */
@SuppressWarnings("unused")
public class MwServiceError implements ServiceError {
    @Nullable private String code;
    @Nullable private String text;
    @Nullable private String html;
    @Nullable private Data data;

    public MwServiceError() {
    }

    public MwServiceError(@Nullable String code, @Nullable String html) {
        this.code = code;
        this.html = html;
    }

    @Override @NonNull public String getTitle() {
        return StringUtils.defaultString(code);
    }

    @SuppressWarnings("checkstyle:magicnumber")
    @Override @NonNull public String getDetails() {
        String retStr = StringUtils.defaultString(html);

        // Special case: if it's a Blocked error, and we get a generic message from the API, then
        // parse the blockinfo structure ourselves.
        if ("blocked".equals(code) && retStr.length() < 100
                && data != null && data.blockinfo != null) {
            retStr = ThrowableUtil.parseBlockedError(data.blockinfo);
        }
        return retStr;
    }

    public boolean badToken() {
        return "badtoken".equals(code);
    }

    public boolean badLoginState() {
        return "assertuserfailed".equals(code);
    }

    public boolean hasMessageName(@NonNull String messageName) {
        if (data != null && data.messages() != null) {
            for (Message msg : data.messages()) {
                if (messageName.equals(msg.name)) {
                    return true;
                }
            }
        }
        return false;
    }

    @Nullable public String getMessageHtml(@NonNull String messageName) {
        if (data != null && data.messages() != null) {
            for (Message msg : data.messages()) {
                if (messageName.equals(msg.name)) {
                    return msg.html();
                }
            }
        }
        return null;
    }

    private static final class Data {
        @Nullable private List<Message> messages;
        @Nullable private BlockInfo blockinfo;

        @Nullable private List<Message> messages() {
            return messages;
        }
    }

    private static final class Message {
        @Nullable private String name;
        @Nullable private String html;

        @NonNull private String html() {
            return StringUtils.defaultString(html);
        }
    }

    public static class BlockInfo {
        private int blockid;
        private int blockedbyid;
        @Nullable private String blockreason;
        @Nullable private String blockedby;
        @Nullable private String blockedtimestamp;
        @Nullable private String blockexpiry;

        public int getBlockId() {
            return blockid;
        }

        @NonNull public String getBlockedBy() {
            return StringUtils.defaultString(blockedby);
        }

        @NonNull public String getBlockReason() {
            return StringUtils.defaultString(blockreason);
        }

        @NonNull public String getBlockTimeStamp() {
            return StringUtils.defaultString(blockedtimestamp);
        }

        @NonNull public String getBlockExpiry() {
            return StringUtils.defaultString(blockexpiry);
        }

        public boolean isBlocked() {
            if (TextUtils.isEmpty(blockexpiry)) {
                return false;
            }
            Date now = new Date();
            Date expiry = DateUtil.iso8601DateParse(blockexpiry);
            return expiry.after(now);
        }
    }
}

File: app/src/main/java/org/wikipedia/dataclient/mwapi/MwServiceError.kt
package org.wikipedia.dataclient.mwapi

import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.wikipedia.dataclient.ServiceError
import org.wikipedia.util.DateUtil
import org.wikipedia.util.StringUtil
import org.wikipedia.util.ThrowableUtil
import org.wikipedia.util.log.L
import java.util.*

@Serializable
class MwServiceError(val code: String? = null,
                     var html: String? = null,
                     val data: Data? = null) : ServiceError {

    fun badToken(): Boolean {
        return "badtoken" == code
    }

    fun badLoginState(): Boolean {
        return "assertuserfailed" == code
    }

    fun hasMessageName(messageName: String): Boolean {
        return data?.messages?.find { it.name == messageName } != null
    }

    fun getMessageHtml(messageName: String): String? {
        return data?.messages?.first { it.name == messageName }?.html
    }

    override val title: String get() = code.orEmpty()

    override val details: String get() = StringUtil.removeStyleTags(html.orEmpty())

    init {
        // Special case: if it's a Blocked error, parse the blockinfo structure ourselves.
        if (("blocked" == code || "autoblocked" == code) && data?.blockinfo != null) {
            runBlocking(CoroutineExceptionHandler { _, t ->
                L.e(t)
            }) {
                html = ThrowableUtil.getBlockMessageHtml(data.blockinfo)
            }
        }
    }

    @Serializable
    class Data(val messages: List<Message>? = null, val blockinfo: BlockInfo? = null)

    @Serializable
    class Message(val name: String?, val html: String = "")

    @Serializable
    open class BlockInfo {

        @SerialName("blockedbyid")
        val blockedById = 0
        @SerialName("blockid")
        val blockId = 0
        @SerialName("blockedby")
        val blockedBy: String = ""
        @SerialName("blockreason")
        val blockReason: String = ""
        @SerialName("blockedtimestamp")
        val blockTimeStamp: String = ""
        @SerialName("blockexpiry")
        val blockExpiry: String = ""

        val isBlocked: Boolean
            get() {
                if (blockExpiry.isEmpty()) {
                    return false
                }
                val now = Date()
                val expiry = DateUtil.iso8601DateParse(blockExpiry)
                return expiry.after(now)
            }
    }
}
